Add a pageSnapping parameter to PageView

jwmcglynn opened the pull request:
This is a change I made to flutter for my application, which enabled me to create a custom scroll bar by introducing a `pageSnapping` property on `PageView`.

When my scroll bar is being dragged, I disable pageSnapping to precisely control the position of the scroll view, and on release I re-enable snapping which results in the nearest page snapping into alignment.

This change includes a test to validate the functionality; let me know if there is anything additional required.

jwmcglynn committed:
Add a pageSnapping parameter to PageView

Setting the pageSnapping property allows extending the PageView scroll
behavior, such as custom scroll animations or custom scroll bars.

Hixie commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -463,6 +470,13 @@ class _PageViewState extends State<PageView> {
+      physics = widget.pageSnapping ? _kPagePhysics : _kNonSnappingPhysics;
+    } else if (widget.pageSnapping) {
+      physics = _kPagePhysics.applyTo(physics);
+    }
```
in the non-snapping, no-custom-physics case, i think you can just pass null rather than having to have _kNonSnappingPhysics.

jwmcglynn replied:
Updated. I also discovered that _kPagePhysics.applyTo(null) is valid, so the line can be further simplified and remove the null check.

Hixie commented on packages/flutter/lib/src/widgets/page_view.dart:
```diff
@@ -502,5 +516,6 @@ class _PageViewState extends State<PageView> {
     description.add(new FlagProperty('reverse', value: widget.reverse, ifTrue: 'reversed'));
     description.add(new DiagnosticsProperty<PageController>('controller', widget.controller, showName: false));
     description.add(new DiagnosticsProperty<ScrollPhysics>('physics', widget.physics, showName: false));
+    description.add(new FlagProperty('pageSnapping', value: widget.pageSnapping, ifFalse: 'noSnapping'));
```
ifFalse: "snapping disabled" or some such.

jwmcglynn replied:
Done

Hixie commented on packages/flutter/test/widgets/page_view_test.dart:
```diff
@@ -368,6 +368,70 @@ void main() {
+      (WidgetTester tester) async {
+    final List<int> log = <int>[];
+
+    Widget build({bool pageSnapping}) {
```
nit: spaces inside the braces

jwmcglynn replied:
Fixed.  This contradicts what "flutter format" does, is that intentional?

Hixie commented:
This looks great. Just some minor nits and then I can land it for you! Thanks!

jwmcglynn commented:
I've responded to feedback and updated the change.  Note that I opted to amend the change to keep this as one commit.  This is my first pull request, I wasn't aware that github would have a hard time diffing it.

If you want I have the original change in a backup branch and can put it back up as two changes.

jwmcglynn committed:
Apply pageSnapping CR feedback

- Remove _kNonSnappingPhysics, use null instead.
- Minor code style fixes.
- It turns out that the forth state is Arkansas, not California.

Hixie commented:
This is great! Thanks for the test, too. I'll make sure this lands once the bot gives it the all clear (it was having issues). Thanks for the contribution!
